### DISCUSSION TITLE
tls connection without verify peer failed

### DISCUSSION THREAD

**Discussion by wiryonolau** [php-amqplib/php-amqplib#1105] [Q&A] [answered]
Hi I want to test tls connection without verifying

Client is php 7.4.33 and Server is Rabbitmq 3.12 both running from docker
without tls everything run flawlessly  

I'm also testing with python since it will communicate with it later.  Using aio-pika, this is work normally
Note that I'm not providing any certificate ( skip verify )
```amqps://principal:principal@localhost:5671/?no_verify_ssl=1```

```php
$amqpConfig = new AMQPConnectionConfig();
$amqpConfig->setHost("localhost");
$amqpConfig->setPort(5671);
$amqpConfig->setUser("principal");
$amqpConfig->setPassword("principal");
$amqpConfig->setIsSecure(true);
$amqpConfig->setSslVerify(false);
$amqpConfig->setSslVerifyName(false);
$amqpConfig->setIoType(AMQPConnectionConfig::IO_TYPE_STREAM);
$amqpConfig->setVhost("/");

$connection = AMQPConnectionFactory::create($amqpConfig);
```

I got this error ``` Server not ready - Invalid frame type 21``` on PHP
and this error in rabbitmq 
```
{unsupported_record_type,65}
TLS server: In state hello at tls_record.erl:564 generated SERVER ALERT: Fatal - Unexpected Message
```

Any idea how to further debug ?

**Comment by lukebakken** [comment]
It appears you have not configured RabbitMQ to use TLS.

If you'd like us to assist please provide a docker compose project that I can clone and run to see the same issue you report, or at the very least, provide _all_ of your RabbitMQ configuration (including certificates) as well as your PHP and aio-pika test code. Thanks.

**Reply by wiryonolau** [reply]
It already run using tls. It works from python. Unless it missing something, for cert I use self sign 

Here is my configuration
part of docker-compose.yml
```yml
...
  rabbitmq:
    image: rabbitmq:3.12-alpine
    ports:
      - 5671:5671
    volumes:
      - ./etc/rabbitmq/cert:/etc/rabbitmq/cert:ro
      - ./etc/rabbitmq/rabbitmq.conf:/etc/rabbitmq/rabbitmq.conf:ro
      - ./etc/rabbitmq/definitions.json:/etc/rabbitmq/definitions.json:ro
    healthcheck:
      test:
        - "CMD"
        - "rabbitmq-diagnostics"
        - "-q"
        - "check_port_connectivity"
      interval: 5s
      timeout: 5s
      retries: 3
    restart: unless-stopped
...
```

rabbitmq.conf
```ini
definitions.import_backend = local_filesystem                                                                                       
definitions.local.path = /etc/rabbitmq/definitions.json

loopback_users.guest = false
listeners.tcp.default = 5672

listeners.ssl.default = 5671
ssl_options.cacertfile = /etc/rabbitmq/cert/rabbitmq.ca.crt
ssl_options.certfile = /etc/rabbitmq/cert/rabbitmq.crt
ssl_options.keyfile = /etc/rabbitmq/cert/rabbitmq.key
ssl_options.verify = verify_none
ssl_options.fail_if_no_peer_cert = false
```

definitions.json
```json
{                                                                                                                                   
  "bindings": [],
  "exchanges": [
    {
      "name": "transaction",
      "vhost": "/",
      "type": "fanout",
      "durable": true,
      "auto_delete": false,
      "internal": false
    }
  ],
  "global_parameters": [],
  "parameters": [],
  "permissions": [
    {
      "configure": ".*",
      "read": ".*",
      "user": "principal",
      "vhost": "/",
      "write": ".*"
    },
  ],
  "policies": [],
  "queues": [],
  "topic_permissions": [],
  "users": [
    {
      "limits": {},
      "name": "principal",
      "password": "principal",
      "tags": []
    }
  ],
  "vhosts": [
    {
      "name": "/"
    }
  ]
}
```

rabbitmq log
```
2023-07-26 23:15:10.838657+00:00 [info] <0.548.0> Ready to start client connection listeners
2023-07-26 23:15:10.843195+00:00 [info] <0.674.0> started TCP listener on [::]:5672
2023-07-26 23:15:10.846215+00:00 [info] <0.694.0> started TLS (SSL) listener on [::]:5671
 completed with 3 plugins.
2023-07-26 23:15:10.900952+00:00 [info] <0.548.0> Server startup complete; 3 plugins started.
2023-07-26 23:15:10.900952+00:00 [info] <0.548.0>  * rabbitmq_prometheus
2023-07-26 23:15:10.900952+00:00 [info] <0.548.0>  * rabbitmq_web_dispatch
2023-07-26 23:15:10.900952+00:00 [info] <0.548.0>  * rabbitmq_management_agent
```

Python way
```python
async def tlspub() -> None:
    # Perform connection                                                                                                            
    constr = "amqps://principal:principal@localhost:5671/?no_verify_ssl=1"
    connection = await connect(constr)

    async with connection:
        # Creating a channel
        channel = await connection.channel()

        logs_exchange = await channel.declare_exchange(
            "transaction", ExchangeType.FANOUT, durable=True
        )

        message_body = b" ".join(
            arg.encode() for arg in sys.argv[2:]
        ) or b"Hello World!"

        message = Message(
            message_body,
            delivery_mode=DeliveryMode.PERSISTENT,
        )

        # Sending the message
        await logs_exchange.publish(message, routing_key="ignore")

        print(f" [x] Sent {message!r}")

```

Rabbitmq log from python
```
2023-07-26 23:40:14.962975+00:00 [info] <0.1390.0> accepting AMQP connection <0.1390.0> (10.230.2.1:54566 -> 10.230.2.3:5671)
2023-07-26 23:40:14.968226+00:00 [info] <0.1390.0> connection <0.1390.0> (10.230.2.1:54566 -> 10.230.2.3:5671): user 'principal' authenticated and granted access to vhost '/'
2023-07-26 23:40:14.977163+00:00 [warning] <0.1390.0> closing AMQP connection <0.1390.0> (10.230.2.1:54566 -> 10.230.2.3:5671, vhost: '/', user: 'principal'):
2023-07-26 23:40:14.977163+00:00 [warning] <0.1390.0> client unexpectedly closed TCP connection
```

If i remove the "s" from the "amqps://" it produce same error as php in rabbitmq log

**Reply by lukebakken** [reply]
Rather than give me _all_ of the code and configuration I need to reproduce your issue, you chose to not give a complete set of reproduction code. I have now taken almost an hour to put together a *COMPLETE* reproduction example here:

https://github.com/lukebakken/php-amqplib-1105

You're lucky I'm feeling generous tonight. In the future, please do NOT take *FREE* software and the time that people put into it for granted, and make it _as easy as possible_ for someone to reproduce an issue that you think you have found. Ideally, all I would have to do is `git clone https://github.com.... && make` to reproduce your issue.

I will finish what I have started and yes, it should be as easy as `git clone https://github.com/lukebakken/php-amqplib-1105.git && make` to set up an environment and hopefully demonstrate your issue.

**Reply by lukebakken** [reply]
Actually, I have spent enough time on reproducing this issue. I would like you, @wiryonolau, to fork my repository ...

https://github.com/lukebakken/php-amqplib-1105

...and open a pull request that finishes the work I have started. The only thing that needs to be done is to make it so that running `make run-aio-pika-repro` and `make run-php-repro` actually do something.

I copied your `aio-pika` code into this file:

https://github.com/lukebakken/php-amqplib-1105/blob/main/aio-pika/repro.py

To run it standalone, you would do this:

```
cd aio-pika
pipenv install --dev
pipenv run repro
```

But, of course, the code you gave me doesn't actually do anything when run as a standalone Python program. I would like you to fix that.

I copied your PHP code to this file:

https://github.com/lukebakken/php-amqplib-1105/blob/main/php-repro/repro.php

...again, as provided, it doesn't do anything.

I will gladly cooperate with you in completing the reproduction code and then investigating the issue, but I'm going to need you to help out. Thanks.

**Reply by wiryonolau** [reply]
Hi you could ask it nicely.

Sometime you can find with just checking the config. 

You can say config seems ok and we need to see all the source code. I'll try built the full source.

Thats why I put it in discussion instead as an issue, someone might already see the problem.

My php repo is at https://github.com/wiryonolau/php-queue

I haven't push the fix and change for tls. I'll update here when it is done.

**Reply by lukebakken** [reply]
> Hi you could ask it nicely

I _am_ being nice. If you think otherwise, spend some time on the LKML.

> Thats wht I put it in discussion instead as an issue,

_I_ converted your original issue, #1104, into this discussion (#1105).

> You can say config seems ok and we need to see all the source code

I already did that, and the code you provided is not runnable.

At first glance, your PHP code looks correct. But, then you put it _entirely_ on me and @ramunasd (the other project maintainer) to take the time to set everything up to see the same issue. It is courteous to do as much of that work as possible yourself so that we can spend time diagnosing and fixing the issue.

This doesn't apply to just `php-amqplib` but _all_ of the vast amount of free software I'm certain you use every day.

**Reply by wiryonolau** [reply, marked as answer]
Hi I need time to prepare my code for share 
Here is what I found.

```php
$config = new AMQPConnectionConfig();

// This doesn't set the correct network protocol to ssl in AMQPConnectionFactory 
// PhpAmqpLib\Wire\IO\StreamIO will still use "tcp" if networkprotocol not set
$config->setIsSecure(true); 
// We need to manually set the network protocol too.
$config->setNetworkProtocol("ssl"); 
```

Still testing ssl 
Setting setSslVerify false works normally as long the cert and ca cert is null

**Reply by wiryonolau** [reply]
Testing using ssl works ok, there is no option to allow_self_signed ssl context thought

**Reply by lukebakken** [reply]
Thank you very much for investigating. We would appreciate a pull request to set the correct network protocol when `setIsSecure(true)` is called.

**Comment by lukebakken** [comment]
@wiryonolau thank you SO MUCH for the pull request - https://github.com/php-amqplib/php-amqplib/pull/1106
